[issue 10989 ] Java Client: KeyValueSchema with AutoConsume component - make it work if the topic schema is still not set

eolivelli opened the pull request:
Fixes #10989

### Motivation

When you use KeyValueSchema together with Schema.AUTO_CONSUME() for the key or for the value, currently (Pulsar 2.8.0) the "consumerBuilder.subscribe()"  method fails because in this case the KeyValueSchema requires to fetch the schema, because one of the BK components uses AutoConsumeSchema, but the schema is not set on the topic.  

### Modifications

- Add some special handling in the subscribe code path in order to deal with KeyValueSchema, the same way we are doing for AutoConsumeSchema
- Add special handling in MessageImpl in order to support schema versioning even in this case, the same way we are doing for AutoConsumeSchema 

### Verifying this change

This change added a case to an existing test.


### Documentation
No need for documentation updates, this is a bug fix

congbobo184 commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java:
```diff
@@ -1012,7 +1013,8 @@ public MemoryLimitController getMemoryLimitController() {
                     if (null == schemaInfo) {
-                        if (!(finalSchema instanceof AutoConsumeSchema)) {
+                        if (!(finalSchema instanceof AutoConsumeSchema)
+                            && !(finalSchema instanceof KeyValueSchema)) {
```
one of keySchema and valueSchema is `AutoConsumeSchema` need this right?

eolivelli replied:
yes, this comes from the fact that above the schema returned `true` to `requireFetchingSchemaInfo`and so it is a dynamic schema

https://github.com/apache/pulsar/blob/0649ef80c5f07fc4a81f909c5c4c79dbef641ead/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L1012

congbobo184 reviewed (approved):
LGTM! left some comment.

eolivelli commented on pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleSchemaTest.java:
```diff
@@ -555,7 +555,16 @@ public void testAutoKeyValueConsume(boolean batching) throws Exception {
+                Schema.KeyValue(
+                        Schema.AUTO_CONSUME(),
+                        Schema.AVRO(V1Data.class),
+                        KeyValueEncodingType.SEPARATED))
```
done

eolivelli commented:
@congbobo184 I have answered to your questions and added the tests for KeyValueEncodingType.INLINE

@codelipenghui I added an assertion that the Consumer<KeyValue<GenericRecord,Pojo>> does not set a schema on the topic if the topic has no schema
